Download genbank file

kdp-cloud opened the pull request:
This PR contains the implementation of downloading a GenBank file of a level 1 construct.

kdp-cloud committed:
Create Genbank file for a level1 Vector download button

- genbank.py: parsing model.Vector to genbank string
- vector.py: function that handles the POST request
- Added a client-side function to download the genbank file.

This is a WIP commit

kdp-cloud committed:
Complete initial implementation of generating Genbank data

- Remove the Bio.Alphabet import. See Following
  [link](https://biopython.org/wiki/Alphabet)
- Added download buttons in table on the client
- Removed download button from constructLevel1View
- Fixed naming conflict: Page types had same name as some Msg types in
  Main.elm
- Fixed imports in Main.elm
- Fixed Links in level0Table and backboneTable:
  - If tables are rendered in Catalgue, links should should load the
    vector in  the AddLevel0Page or AddBackbonePage.
  - If tables are rendered in AddLevel1Page, links should add the
    level0 or backbone to the level1 construct.

Limitations:
When downloading a Genbank file, the output is all rendered on 1
line. '\n' is not recognised as a new line character.

kdp-cloud committed:
Ask git to store LF only to mark the end-of-line

This is because Windows marks end-of-line with CRLF and we don't want
to be poluting git history with end-of-line marking changes.

kdp-cloud committed:
Implement digestion

Digestion for:
- Level 0
- Backbones

This happens prior to initial data import and also adding new vectors
from the web client.

kdp-cloud committed:
Simplify vector catalogue in client

When switching between pages, the overhangRadioRow in the
vectorcatalogue should always show all possible overhangs and not a
subset.

- Download Genbank by Vector `id`
- Fixied the overhangShape in the vectorCatalogue
- Removed redundant Msg case from update function (after rebase)
- Revert saving a `Maybe Vector`

MaybeJustJames commented on client/src/Api.elm:
```diff
@@ -21,10 +21,11 @@ type alias Api msg =
     , authorize : AuthCode -> Cmd msg
     , vectors : Auth -> Cmd msg
-    , save : Auth -> Vector -> Cmd msg
+    , save : Auth -> Maybe Vector -> Cmd msg
```
What!? What is this for? Why are we saving a `Maybe Vector`?

MaybeJustJames replied:
```suggestion
    , save : Auth -> Vector -> Cmd msg
```

MaybeJustJames replied:
I've fixed this myself

MaybeJustJames commented on client/src/Api.elm:
```diff
@@ -111,15 +123,15 @@ mkVectors url expect auth =
 
 
-mkSave : ApiUrl -> Expect msg -> Auth -> Vector -> Cmd msg
+mkSave : ApiUrl -> Expect msg -> Auth -> Maybe Vector -> Cmd msg
```
```suggestion
mkSave : ApiUrl -> Expect msg -> Auth -> Vector -> Cmd msg
```

MaybeJustJames commented on client/src/Api.elm:
```diff
@@ -111,15 +123,15 @@ mkVectors url expect auth =
                 apiUrl =
                     case vec of
-                        LevelNVec _ ->
+                        Just (LevelNVec _) ->
```
```suggestion
                        LevelNVec _ ->
```

MaybeJustJames commented on client/src/Molecules.elm:
```diff
@@ -390,23 +390,28 @@ vectorDecoder =
 
 
-vectorEncoder : Vector -> Encode.Value
+vectorEncoder : Maybe Vector -> Encode.Value
```
```suggestion
vectorEncoder : Vector -> Encode.Value
```

MaybeJustJames replied:
I've fixed this myself

MaybeJustJames commented on client/src/Molecules.elm:
```diff
@@ -390,23 +390,28 @@ vectorDecoder =
 vectorEncoder vector =
     case vector of
-        Level0Vec vec ->
+        Just (Level0Vec vec) ->
```
```suggestion
        Level0Vec vec ->
```

MaybeJustJames commented on client/src/Molecules.elm:
```diff
@@ -390,23 +390,28 @@ vectorDecoder =
             level0Encoder vec
 
-        BackboneVec vec ->
+        Just (BackboneVec vec) ->
```
```suggestion
        BackboneVec vec ->
```

MaybeJustJames commented on client/src/Molecules.elm:
```diff
@@ -390,23 +390,28 @@ vectorDecoder =
             backboneEncoder vec
 
-        LevelNVec vec ->
+        Just (LevelNVec vec) ->
```
```suggestion
        LevelNVec vec ->
```

MaybeJustJames commented on client/src/Molecules.elm:
```diff
@@ -390,23 +390,28 @@ vectorDecoder =
+        Just (LevelNVec vec) ->
             levelNEncoder vec
 
+        _ ->
+            backboneEncoder initBackbone
+
```
```suggestion
```

MaybeJustJames commented on client/src/Molecules.elm:
```diff
@@ -390,23 +390,28 @@ vectorDecoder =
+
 
 backboneEncoder : Backbone -> Encode.Value
 backboneEncoder backbone =
     Encode.object
-        [ ( "name", Encode.string backbone.name )
+        [ ( "id", Encode.int backbone.id )
+        , ( "sequence_length", Encode.int backbone.sequenceLength )
+        , ( "name", Encode.string backbone.name )
```
I don't think this is correct. See `server/app/schemas.py:VectorIn`. No `id`, no `sequence_length`.

kdp-cloud replied:
Indeed, we don't need them for submitting vectors

MaybeJustJames commented on client/src/Molecules.elm:
```diff
@@ -564,8 +573,6 @@ levelNEncoder level1 =
         , ( "level"
           , Encode.int 3
           )
-        , ( "inserts", Encode.list (\insert -> level0Encoder insert) level1.inserts )
-        , ( "backbone", backboneEncoder (Maybe.withDefault initBackbone level1.backbone) )
```
Indeed, but you do need a `children` key?

kdp-cloud replied:
I don't really follow... schemas.py: VectorIn has a field `children`.
So I assume I need to pass it to the server.

MaybeJustJames replied:
Indeed, but there is no `children` key in your encoder.

kdp-cloud replied:
What I meant to say was that `levelNEncoder` does have a `children` key on line 580:

```elm
         ( "children"
          , Encode.list Encode.int
                ((toList level1.backbone
                    |> List.map .id
                 )
                    ++ List.map .id level1.inserts
                )
          )

```

Is this OK? Or do you mean something else?

MaybeJustJames replied:
Indeed. I missed this. Thanks

MaybeJustJames commented on client/src/Router.elm:
```diff
@@ -24,9 +24,9 @@ import Url exposing (Protocol(..), Url)
-    | AddLevel1
-    | AddLevel0
-    | AddBackbone
+    | AddLevel1Page
+    | AddLevel0Page
+    | AddBackbonePage
```
Why? This change makes the naming inconsistent. I strongly recommend undoing this change

kdp-cloud replied:
Main.elm already has those as type Msg. I did it to avoid confusion.

MaybeJustJames replied:
Ah ok. Fair enough. Can you make the naming consistent too then: `LoginPage`, `CataloguePage`, and `AdminPage`..

kdp-cloud replied:
Done!

MaybeJustJames commented on server/app/genbank.py:
```diff
@@ -11,6 +18,9 @@
 )
 
+from app import model, deps
+from app.level import VectorLevel
```
This is imported twice now.
```suggestion
```

kdp-cloud replied:
I've deleted line 22.

MaybeJustJames commented on server/app/genbank.py:
```diff
@@ -56,7 +66,7 @@ def my_filter(feature: Any) -> bool:
 
 
-# Function that reads in a genbank file and converts it into a json
+# Function that reads in a genbank file and converts it into a GenBankData object
```
Should be a docstring. Don't need to name types because they're in the annotations.

kdp-cloud replied:
I think I did it already:
```python
def convert_gbk_to_vector(genbank_file, level: VectorLevel) -> GenbankData:
    """
    Function that reads in a genbank file and converts it into a GenBankData object.
    """
```

MaybeJustJames commented on server/app/genbank.py:
```diff
@@ -102,3 +112,96 @@ def convert_gbk_to_vector(genbank_file, level: VectorLevel) -> GenbankData:
     )
+
+
+# Converts a model.Vector to a Genbank output
```
Should be a docstring. Be a little more clear (Is it a levelN or any Vector? Why only LevelN? Are you serialising to GenBank format in a string or is the output a filename? etc)
* I would prefer a name like `serialise_to_genbank()`

kdp-cloud replied:
I've added the docstring and changed the name to "serialize_to_genbank".

Strictly speaking you could use this function for generating a genbank file of any kind of vector. I'm not sure if it makes sense to generate a genbank file of a level 0 vector  or a backbone vector, since the user probably already has it (when you submit a level 0 or backbone, you provide a genbank file).

From that perspective it made sense to use "levelN" in the function name...

MaybeJustJames commented on server/app/genbank.py:
```diff
@@ -102,3 +112,96 @@ def convert_gbk_to_vector(genbank_file, level: VectorLevel) -> GenbankData:
+    vector_record.features = feature_list
+
+    # Writing record content to an in-memory file and returning its content.
+    outf = io.StringIO()
+    SeqIO.write(vector_record, outf, "genbank")
+    content = outf.getvalue()
+    outf.close()
+    return content
```
Turns out this can be done with a context manager:
```suggestion
    with io.StringIO() as outf:
        SeqIO.write(vector_record, outf, "genbank")
        return outf.getvalue()
```

MaybeJustJames commented on server/app/genbank.py:
```diff
@@ -102,3 +112,96 @@ def convert_gbk_to_vector(genbank_file, level: VectorLevel) -> GenbankData:
+    return content
+
+
+class mod_Reference(Bio.SeqFeature.Reference):
+    """
+    This class extends the existing Record.Reference class.
+    """
```
Why is this even being defined? [The documentation](https://biopython.org/docs/1.76/api/Bio.SeqFeature.html#Bio.SeqFeature.Reference) states that there are already `authors`, `title`, and `journal` attributes.

Also, the journal is being set to `Generated $TIME`?!

kdp-cloud replied:
Indeed but you can't set the attribute when initializing a `SeqFeature.Reference`. It seemed neater that way but feel free to change it...

In every genbank file I got from Delphine, the `journal` attribute said something like this: 

>JOURNAL   Exported Tuesday, Mar 21, 2017 from SnapGene 3.1.4

So it seemed nice if the GG2 app did something similar.

kdp-cloud replied:
The mod_Reference class has been replaced by a helper function `make_reference` that does the same thing.

MaybeJustJames commented on server/app/vectors.py:
```diff
@@ -162,3 +163,40 @@ def add_leveln(
+def get_level1_genbank(
+    vector_id: int,
+    database: Session = Depends(deps.get_db),
+    # current_user: schemas.User = Depends(deps.get_current_user),
```
```suggestion
    current_user: schemas.User = Depends(deps.get_current_user),
```

kdp-cloud replied:
current_user isn't used in this function.

MaybeJustJames replied:
Yes. In general you can just prefix with an underscore. But this should raise some alarm bells. Without authentication, anyone can download any vector! Without checking if `current_user` is authorised then any logged in user can download any vector. Better check they're allowed access first.

kdp-cloud replied:
I agree that it's best to check for authorization but not with `current_user`.

E.g.: If user x (= `current_user`) makes a level 1 construct, using level 0 elements and a backbone of user y and z, he will not be able to generate a GenBank file of his level 1 construct.

MaybeJustJames replied:
I don't really follow your reasoning. If `x` dies not have access to those backbones and level 0s how can they build the construct in the first place. Let alone download a genbank

MaybeJustJames commented on server/app/vectors.py:
```diff
@@ -162,3 +163,40 @@ def add_leveln(
+    # current_user: schemas.User = Depends(deps.get_current_user),
+) -> str:
+    """
+    This function handles a POST request from the UI
+    for getting a genbank file of a submitted Level 1.
+
+    - Accepts schemas.LevelNToAdd as input
+    - Queries the database for the model.Vector object
+    - Parses the model.Vector object to a genbank file
+    - Returns a genbank file as a string
```
```suggestion
    This function handles a request for a Vector to
    be serialised to GenBank format.
```

kdp-cloud replied:
Done!

MaybeJustJames reviewed (changes requested):
I think a lot of mistakes because you're under the weather. Please don't feel bad, this happens... you're human ;)

MaybeJustJames commented on server/app/genbank.py:
```diff
@@ -102,3 +112,96 @@ def convert_gbk_to_vector(genbank_file, level: VectorLevel) -> GenbankData:
+
+# Converts a model.Vector to a Genbank output
+def convert_LevelN_to_genbank(
+    vector: model.Vector, database: Session = Depends(deps.get_db)
```
You're not using your `database` parameter.
```suggestion
    vector: model.Vector
```

MaybeJustJames commented on client/src/Main.elm:
```diff
@@ -257,85 +272,104 @@ view model =
         , row [ spacing 20 ]
             [ button_ (Just ToggleAll) "Toggle all"
             ]
-        , Accordion.accordion
```
Revert this please. Style the `Accordion` in `Accordion.elm`

kdp-cloud replied:
Done!

MaybeJustJames commented on client/src/Main.elm:
```diff
@@ -113,6 +114,13 @@ changePage router url =
             Router (Router.changePage rtr url)
 
 
+getRouterPage : Router -> Page
+getRouterPage router =
```
```suggestion
routerPage : Router -> Page
routerPage router =
```

kdp-cloud replied:
Done!

MaybeJustJames commented on client/src/Main.elm:
```diff
@@ -964,12 +990,17 @@ overhangRadioRow model =
-            case model.vectorToAdd of
-                Just (LevelNVec _) ->
-                    overhangShape model.currLevel1App
+            case getRouterPage model.router of
```
```suggestion
            case routerPage model.router of
```

kdp-cloud replied:
Done!

MaybeJustJames commented on client/src/Main.elm:
```diff
@@ -1027,9 +1058,17 @@ level0Table model =
-            case model.vectorToAdd of
-                Just (LevelNVec level1) ->
-                    buttonLink_ (Just (UpdateVectorToAdd <| Just <| LevelNVec <| { level1 | inserts = appendInsertToLevel1 level1.inserts level0 })) level0.name
+            case getRouterPage model.router of
```
```suggestion
            case routerPage model.router of
```

kdp-cloud replied:
Done!

MaybeJustJames commented on client/src/Main.elm:
```diff
@@ -1307,7 +1356,12 @@ update msg model =
-            ( model, gotoRoute model page )
+            case page of
+                Catalogue ->
+                    ( { model | vectorToAdd = Just <| Level0Vec initLevel0 }, gotoRoute model page )
```
Huh!?

kdp-cloud replied:
Yeah not ideal...

The issue is the `overhangRadioRow`. If vectorToAdd is a Level0Vec, `theOverhangShape` = `allBsa1Overhangs`. If vectorToAdd is a LevelNVec, `overhangShape` depends on `currLevel1App`.

So when you switch from the "Add Level 1 page" to the vector catalog, you should be able to see all kinds of overhangs and not a subset.

MaybeJustJames commented on client/src/Main.elm:
```diff
@@ -1419,7 +1473,38 @@ update msg model =
+            ( model, model.api.save model.auth <| LevelNVec level1 )
+
+        DownloadGenbankFile level1 ->
+            ( { model | vectorToAdd = Just (LevelNVec level1) }, model.api.genbank model.auth level1.id )
```
Why is the `model` being changed here?

kdp-cloud replied:
vectorToAdd is by default Nothing. So if you click the download button, you'll try to change the content of "Nothing".

MaybeJustJames commented on client/src/Main.elm:
```diff
@@ -1517,6 +1622,11 @@ appendInsertToLevel1 l0List newL0 =
 
+saveGenbank : String -> String -> Cmd Msg
+saveGenbank vectorName content =
+    File.Download.string (String.append vectorName ".gbk") "text/genbank" content
```
MIME type for Genbank seems to be:
* `chemical/x-genbank`
  * https://www.ch.ic.ac.uk/chemime/
  * https://bugs.freedesktop.org/show_bug.cgi?id=335
  * https://mimeapplication.net/chemical-x-genbank
  * https://en.wikipedia.org/wiki/Chemical_file_format#The_Chemical_MIME_Project
  * https://github.com/dleidert/chemical-mime/blob/master/src/chemical-mime-database.xml.in#L736=

MaybeJustJames replied:
Also note the extension used, `.gb` or `.gen`

kdp-cloud committed:
Added electronic traineeship notebook

kdp-cloud committed:
Authenticate user when downloading a Genbank file

Also:
- Replaced mod_Reference class by make_reference function
- Removed id and sequenceLength from backboneEncoder.
- Deleted double import

Fix Vector encoders:
- Do not encode `id`
- Uniform way of encoding default `null` or ""
- Remove single use of simple function `isBsmb1FreeToBool`

MaybeJustJames reviewed (approved):
Great work @kdp-cloud. Sorry this took so long to merge.
